tests/tsl/tsl3/test.cs

```diff
@@ -0,0 +1,268 @@
+                    Assert.Throws<InvalidOperationException>(() => reader.Nums[1] = 10000);
+                    Assert.Throws<InvalidOperationException>(() => reader.Nums.Add(100000));
+                }
+            }
+        }
+
+        [Fact]
+        public void Reader_ShouldHaveProperlyTypedProperties()
+        {
+            {
+                var type = typeof(ReqReader);
+                var properties = type.GetMembers().Where(m => m.MemberType == MemberTypes.Property).Cast<PropertyInfo>().ToList();
+                Assert.True(properties.Single(m => m.Name == "FieldBeforeList").PropertyType == typeof(int), "FieldBeforeList");
+                Assert.True(properties.Single(m => m.Name == "Nums").PropertyType == typeof(intListAccessor), "Nums");
+                Assert.True(properties.Single(m => m.Name == "FieldAfterList").PropertyType == typeof(byte), "FieldAfterList");
+            }
+            {
+                var type = typeof(RespReader);
+                var property = (PropertyInfo) type.GetMember("result").Single();
+                Assert.True(property.PropertyType == typeof(StringAccessor));
+            }
+        }
+
+    }
+
+    [Collection("TestServer Collection")]
+    public unsafe class ProtocolTest
+    {
+        private TrinityServerFixture Fixture { get; }
+        public ProtocolTest(TrinityServerFixture fixture) { Fixture = fixture; }
+
+        [Theory]
+        [InlineData(new byte[] {1, 2, 3, 4})]
+        [InlineData(new byte[] {2, 0, 4, 8})]
+        [InlineData(new byte[] {123, 124, 75, 43})]
+        [InlineData(new byte[] {77, 88, 9, 8})]
+        [InlineData(new byte[] {77, 88, 9, 8, 77, 88, 9, 8, 77, 88, 9, 8, 77, 88, 9, 8, })]
+        public void SynWithRsp_Test(byte[] nums)
+        {
+            Assert.True(nums.Length >= 4);
+            var numList = nums.ToList();
+            using (var writer = new ReqWriter())
+            {
+                writer.FieldBeforeList = numList[0];
+                writer.Nums.Add(numList[1]);
+                Console.WriteLine(string.Join(" ", numList.Skip(2).Select(Convert.ToInt32).ToList()));
+                writer.Nums.AddRange(numList.Skip(2).Select(Convert.ToInt32).ToList());
+                writer.Nums.RemoveAt(writer.Nums.Count - 1);
+                writer.FieldAfterList = numList.Last();
+                using (var response = Global.CloudStorage.TestSynWithRspToTestServer(Global.MyServerID, writer))
+                {
+                    Assert.Equal(string.Join(" ", numList), response.result);
+                }
+            }
+        }
+
+        [Fact]
+        public void Syn_Test()
+        {
+            using (var writer = new ReqWriter())
+            {
+                writer.FieldBeforeList = 2;
+                writer.Nums.Add(0);
+                writer.Nums.Add(7);
+                writer.FieldAfterList = 3;
+                Assert.Throws<IOException>(() => Global.CloudStorage.TestSynToTestServer(Global.MyServerID, writer));
```

Comment:
we'd better do some checking by storing something into the local memory storage in the handler (then check it in the test) rather than relying on IOExceptions, because the exceptions do not carry any metadata.